File: play-services-location/core/src/main/java/org/microg/gms/location/ReportingAndroidService.java
/*
 * Copyright (C) 2013-2017 microG Project Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.microg.gms.location;

import android.os.RemoteException;

import com.google.android.gms.common.internal.ConnectionInfo;
import com.google.android.gms.common.internal.GetServiceRequest;
import com.google.android.gms.common.internal.IGmsCallbacks;

import org.microg.gms.BaseService;
import org.microg.gms.common.GmsService;

public class ReportingAndroidService extends BaseService {
    private ReportingServiceImpl reportingService = new ReportingServiceImpl(this);

    public ReportingAndroidService() {
        super("GmsLocReportingSvc", GmsService.LOCATION_REPORTING);
    }

    @Override
    public void handleServiceRequest(IGmsCallbacks callback, GetServiceRequest request, GmsService service) throws RemoteException {
        ConnectionInfo info = new ConnectionInfo();
        info.features = GoogleLocationManagerService.FEATURES;
        callback.onPostInitCompleteWithConnectionInfo(0, reportingService.asBinder(), info);
    }
}

File: play-services-location/core/src/main/kotlin/org/microg/gms/location/reporting/ReportingAndroidService.kt
/*
 * SPDX-FileCopyrightText: 2023 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */
package org.microg.gms.location.reporting

import android.os.RemoteException
import android.util.Log
import androidx.lifecycle.lifecycleScope
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.internal.ConnectionInfo
import com.google.android.gms.common.internal.GetServiceRequest
import com.google.android.gms.common.internal.IGmsCallbacks
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.microg.gms.BaseService
import org.microg.gms.common.GmsService
import org.microg.gms.common.GooglePackagePermission
import org.microg.gms.common.PackageUtils
import org.microg.gms.location.manager.FEATURES

class ReportingAndroidService : BaseService("GmsLocReportingSvc", GmsService.REPORTING) {
    @Throws(RemoteException::class)
    override fun handleServiceRequest(callback: IGmsCallbacks, request: GetServiceRequest, service: GmsService) {
        val packageName = PackageUtils.getAndCheckCallingPackage(this, request.packageName)
            ?: throw IllegalArgumentException("Missing package name")
        val canSynchronizeSettings = PackageUtils.callerHasGooglePackagePermission(
            this,
            GooglePackagePermission.REPORTING
        )
        if (canSynchronizeSettings) {
            lifecycleScope.launch(Dispatchers.IO) {
                runCatching {
                    synchronizePendingAccountOptIns(this@ReportingAndroidService)
                }.onFailure {
                    Log.w(TAG, "Failed to synchronize pending account opt-ins", it)
                }
            }
        }
        callback.onPostInitCompleteWithConnectionInfo(
            CommonStatusCodes.SUCCESS,
            ReportingServiceInstance(this, packageName, lifecycle),
            ConnectionInfo().apply { features = FEATURES }
        )
    }
}
